projects/charger/inc/generic_can.h

```diff
@@ -0,0 +1,88 @@
+  // Doesn't support ACKable messages (defaults to enabled).
+  StatusCode (*register_rx)(struct GenericCan *can, GenericCanRx rx_handler, uint32_t id,
+                            void *context);
+  StatusCode (*enable_rx)(struct GenericCan *can, uint32_t id);
+  StatusCode (*disable_rx)(struct GenericCan *can, uint32_t id);
+} GenericCanInterface;
+
+typedef struct GenericCan {
+  GenericCanInterface *interface;
+  GenericCanRxStorage rx_storage[NUM_GENERIC_CAN_RX_HANDLERS];
+} GenericCan;
+
+// Usage:
+//
+// GenericCan uses dynamic dispatch where each struct for an implementation variant of CAN carries
+// with it an implementation of tx, register_rx, enable_rx and disable_rx. To use a given
+// implementation one simply needs to pass the specific type of GenericCan<Network|Hw|Uart> to one
+// of the following functions. Since all three of those types use GenericCan as a base by casting to
+// (GenericCan *). It is possible to derive the interface function pointers without knowing about
+// the rest of the data carried in the struct. This allows an object oriented approach where the
+// remaining variable for a given implementation are "private/protected" meanwhile the interface is
+// consistent for all variants of GenericCan.
+//
+// Example:
+//
+// GenericCanUart can_uart = { 0 };
+// generic_can_uart_init(&can_uart, UART_PORT_1);
+//
+// // Transmit
+// GenericCanMsg my_msg = { 0 };
+//
+// // Populate...
+//
+// generic_can_tx((GenericCan *)&can_uart, &my_msg);
+//
+// The primary advantage of this is that if you build a function that relies on CAN it can accept
+// any GenericCan implementation and be ignorant to its underlying behavior. For example:
+//
+// static uint64_t s_data;
+//
+// void send_internal_data(const GenericCan *can) {
+//   GenericCanMsg my_msg = { .data = &s_data };
+//   generic_can_tx(&can, &my_msg);
+// }
+//
+// In this way the data can be sent via UART, CAN Hw or CAN Network without the caller knowing which
+// specific CAN variant it is dealing with!
+
+// Transmits a GenericCanMsg.
+//
+// Note: if backed by CAN Network |msg| needs to be compatible with CANMessage and its |source_id|
+// will automatically be set to the ID of the broadcaster as determined in |device_id| of a
+// CANSettings struct. Also |type| must be CAN_MSG_TYPE_DATA as this TX mode does not support acks
+// (therefore its CANMessage.msg_id must be in the range [0, 2047].
+StatusCode generic_can_tx(const GenericCan *can, const GenericCanMsg *msg);
+
+// Registers a |rx_handler| (enabled by default) to a |raw_id|.
+StatusCode generic_can_register_rx(GenericCan *can, GenericCanRx rx_handler, uint32_t raw_id,
+                                   void *context);
+
+// Enables the stored GenericCanRx for |raw_id|.
+StatusCode generic_can_enable_rx(GenericCan *can, uint32_t raw_id);
+
+// Disables the stored GenericCanRx for |raw_id|.
+StatusCode generic_can_disable_rx(GenericCan *can, uint32_t raw_id);
```

Comment:
Is disabling something we need to support? I think it might simplify some of your code if you don't need to check for it.

Reply:
I'd like to disable it so that the permission handlers on charger don't need to be aware whether the charger is in a valid state, rather the handler is made inactive I can do this with a context pointer but this interface is cleaner IMO. I understand it would simplify things but I don't think the cost is that high.

Reply:
Maybe I'm misunderstanding something, but why do the permission handlers ever need to be aware what state the charger is in? Doesn't the charger just output CAN messages about its state that are more for telemetry?

Reply:
That's what is happening more or less. The charger sends a message saying it is connected and asking if it can charge. Power distribution responds with an answer. Power distribution will attempt to enter charging if it is in idle as well hence why the message isn't a simple direct reply. The idea for the permissions handler is that it receives whether power distribution has given permission based on whether it is valid to attempt to charge. Meanwhile, independently, the charger board is monitoring the state of the charger circuit over CAN. If something goes wrong on the charger I want the permissions handler to stop handling the RX since receiving permissions is then invalid.

Reply:
I mean instead of having a request/response sequence, why not change it to a notify/command? The messaging is the same, but it moves responsibility into power distribution and removes the need for the charger to be aware of "plugged in but not charging" or similar. All it needs to worry about is whether the relay is closed or open.

Reply:
Ok re: rx I still don't see why we need to remove the option though? It hardly adds any complexity.

Wait, what? I'm so confused... According to the charger spec all you have to do is supply it a CAN message every second to get it to supply power. Are we talking about the same [part](https://uwmidsun.atlassian.net/wiki/spaces/ELEC/pages/92635386/Charger+and+Required+Parts?preview=/92635386/105611288/Elcon%20CAN%20Specification%20(1).pdf)??

Reply:
It's more that having the ability to enable/disable RX is a bit of an odd paradigm vs. setting filters on IDs. I think it would make sense to try to keep the naming consistent with the existing CAN frameworks.

I'm referring to the charger interface. If power distribution is responsible for opening/closing the relay, the charger interface doesn't need to have the concept of "permissions". It just needs to notify power distribution that the charger is powered. My understanding of your situation is that you disable charger-side CAN network RX, but I don't see the benefit of doing that. Even if we aren't charging, what's the benefit of actually ignoring the charger's CAN messages?

RE: charger spec. Yes, that's what I meant by outputs the requested DC voltage. In fact, I don't think the charger's CAN controller will even be on if the charger isn't plugged into AC power. It seems to be designed to output 12V, not take it as input.

I just wanted to make sure you knew there was a distinction between the J1772 interface circuit that is used to detect and setup the charging station and the charger in the car.

We can use the interface circuit to detect that charging can occur, then notify power distribution. Power distribution can then choose whether to close the charger relay. The idea is that it's easier to move responsibility to a central point (power distribution) and keep the CAN interfaces simple. Since we already have failure-tolerant relay control code, it would make sense to reuse it.